File: surface/widgets/compose/components/TypographyText.kt
// ============================================================
// Clef Surface Compose Widget — TypographyText
//
// Renders text using Clef Surface typography scale tokens.
// Maps token names (display, heading, body, label, caption)
// to Material 3 typography styles.
// ============================================================

package clef.surface.compose.components

import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextOverflow

// --------------- Component ---------------

@Composable
fun TypographyText(
    text: String,
    variant: String = "body",
    size: String = "md",
    weight: String? = null,
    color: Color = Color.Unspecified,
    align: TextAlign? = null,
    maxLines: Int = Int.MAX_VALUE,
    overflow: TextOverflow = TextOverflow.Clip,
    modifier: Modifier = Modifier,
) {
    val style = resolveTypographyStyle(variant, size)
    val fontWeight = when (weight) {
        "bold" -> FontWeight.Bold
        "semibold" -> FontWeight.SemiBold
        "medium" -> FontWeight.Medium
        "light" -> FontWeight.Light
        else -> style.fontWeight
    }

    Text(
        text = text,
        style = style.copy(fontWeight = fontWeight),
        color = color,
        textAlign = align,
        maxLines = maxLines,
        overflow = overflow,
        modifier = modifier,
    )
}

@Composable
private fun resolveTypographyStyle(variant: String, size: String): TextStyle {
    return when (variant) {
        "display" -> when (size) {
            "lg" -> MaterialTheme.typography.displayLarge
            "md" -> MaterialTheme.typography.displayMedium
            "sm" -> MaterialTheme.typography.displaySmall
            else -> MaterialTheme.typography.displayMedium
        }
        "heading" -> when (size) {
            "lg" -> MaterialTheme.typography.headlineLarge
            "md" -> MaterialTheme.typography.headlineMedium
            "sm" -> MaterialTheme.typography.headlineSmall
            else -> MaterialTheme.typography.headlineMedium
        }
        "title" -> when (size) {
            "lg" -> MaterialTheme.typography.titleLarge
            "md" -> MaterialTheme.typography.titleMedium
            "sm" -> MaterialTheme.typography.titleSmall
            else -> MaterialTheme.typography.titleMedium
        }
        "body" -> when (size) {
            "lg" -> MaterialTheme.typography.bodyLarge
            "md" -> MaterialTheme.typography.bodyMedium
            "sm" -> MaterialTheme.typography.bodySmall
            else -> MaterialTheme.typography.bodyMedium
        }
        "label" -> when (size) {
            "lg" -> MaterialTheme.typography.labelLarge
            "md" -> MaterialTheme.typography.labelMedium
            "sm" -> MaterialTheme.typography.labelSmall
            else -> MaterialTheme.typography.labelMedium
        }
        "caption" -> MaterialTheme.typography.labelSmall
        else -> MaterialTheme.typography.bodyMedium
    }
}

File: surface/widgets/nativescript/components/TypographyText.ts
// ============================================================
// Clef Surface NativeScript Widget — TypographyText
//
// Renders text using Clef Surface typography scale tokens.
// Maps token names (display, heading, body, label, caption)
// to NativeScript font sizes and weights.
// ============================================================

import { Label } from '@nativescript/core';

// --------------- Typography Scale ---------------

const FONT_SIZES: Record<string, Record<string, number>> = {
  display: { lg: 57, md: 45, sm: 36 },
  heading: { lg: 32, md: 28, sm: 24 },
  title: { lg: 22, md: 16, sm: 14 },
  body: { lg: 16, md: 14, sm: 12 },
  label: { lg: 14, md: 12, sm: 11 },
  caption: { lg: 11, md: 11, sm: 11 },
};

// --------------- Props ---------------

export interface TypographyTextProps {
  text: string;
  variant?: string;
  size?: string;
  weight?: string;
  color?: string;
  align?: 'left' | 'center' | 'right';
  maxLines?: number;
}

// --------------- Component ---------------

export function createTypographyText(props: TypographyTextProps): Label {
  const {
    text,
    variant = 'body',
    size = 'md',
    weight,
    color,
    align,
    maxLines,
  } = props;

  const label = new Label();
  label.text = text;
  label.className = `clef-typography clef-typography-${variant}-${size}`;

  const fontSize = FONT_SIZES[variant]?.[size] ?? FONT_SIZES.body.md;
  label.fontSize = fontSize;

  const fontWeight = weight ?? (variant === 'display' || variant === 'heading' ? 'bold' : 'normal');
  label.fontWeight = fontWeight as any;

  if (color) label.color = color as any;
  if (align) label.textAlignment = align;
  if (maxLines) label.maxLines = maxLines;

  return label;
}

createTypographyText.displayName = 'TypographyText';
export default createTypographyText;
